IBM Granite

gabe-l-hart opened the pull request:
## Description

This PR is intended to expand the functionality of the researcher to work well with the [IBM Granite](https://ollama.com/library/granite3.2) model family. To do so, it makes two key architectural changes:

* Create the `PromptFamily` abstraction in `prompts.py` so that prompting can be customized to match expected model formats
* Add the ability to set `LLM_KWARGS` and `EMBEDDING_KWARGS` via config when running with `ollama`

## Changes

* Add abstraction for `PromptFamily`:
  * Move all free functions from `prompts.py` into the `PromptFamily` class
  * Add the `PromptFamily` enum in `enum.py`
  * Add the `get_prompt_family` factory function to instantiate the right prompt family based on the enum or string from the `PromptFamily` enum.
  * Plumb `PROMPT_FAMILY` through the `variables` module
  * Use the configured `prompt_family` in all places where a prompt is needed
* Add support for `LLM_KWARGS` and `EMBEDDING_KWARGS`:
  * Add both to `variables`
  * Add dict parsing with `json` to `config.py`

## Testing

With these changes, I wrote the following simple script to run against the Granite family:

**granite.py**

```py
import asyncio
import json
import os
import sys
from gpt_researcher import GPTResearcher

os.environ["FAST_LLM"] = "ollama:granite3.2:2b"
os.environ["SMART_LLM"] = "ollama:granite3.2:8b"
os.environ["STRATEGIC_LLM"] = "ollama:granite3.2:8b"
os.environ["EMBEDDING"] = "ollama:granite-embedding:278m"
os.environ["OPENAI_API_KEY"] = "ollama"
os.environ["OLLAMA_BASE_URL"] = "http://localhost:11434"
os.environ["RETRIEVER"] = "duckduckgo"
os.environ["LLM_KWARGS"] = json.dumps({"num_ctx": 1024 * 128})

if not (query := " ".join(sys.argv[1:])):
    query = input("?> ")

researcher = GPTResearcher(
    query=query,
    report_type="research_report",
    prompt_family="granite",
)

loop = asyncio.get_event_loop()
loop.run_until_complete(researcher.conduct_research())
loop.run_until_complete(researcher.write_report())
```

Here are some examples of running with Granite:

<details>
<summary>Research best practices for optimizing cloud computing costs for large enterprises. Compare cost-saving strategies across major cloud service providers, including pricing models, workload optimization techniques, and case studies of successful cost reduction.</summary>

# Cloud Cost Optimization Best Practices for Large Enterprises: A Comparative Analysis of Major Providers



## Executive Summary



This report examines best practices for optimizing cloud computing costs in large enterprises, focusing on pricing models, workload optimization techniques, and case studies across major cloud service providers (CSPs): Amazon Web Services (AWS), Microsoft Azure, and Google Cloud Platform (GCP). By implementing these strategies, organizations can significantly reduce expenses while maintaining performance, scalability, and security.



## 1. Introduction



Cloud computing has revolutionized the IT landscape by offering scalable, flexible, and cost-effective solutions for businesses (KnowledgeHut, 2024). However, managing cloud costs remains a challenge for large enterprises due to the complexity of pricing models and the potential for overspending. This report explores best practices for optimizing cloud computing costs across major CSPs, emphasizing cost-saving strategies, workload optimization techniques, and real-world case studies.



## 2. Pricing Models and Cost-Saving Strategies



### 2.1 On-Demand vs. Reserved Instances (RIs)



On-demand pricing offers flexibility but can lead to higher costs due to the lack of commitment. Reserved Instances, on the other hand, provide substantial discounts in exchange for a one- or three-year commitment (Lucidity Cloud, 2025). For instance, AWS RIs offer up to 75% discount compared to on-demand pricing (AWS, n.d.). Azure and GCP also provide similar Reserved Instance models (Microsoft, n.d.; Google, n.d.).



### 2.2 Savings Plans



Savings Plans, introduced by AWS, offer greater flexibility than RIs by covering a broader range of services and instance families. By committing to a consistent amount of usage measured in dollars per hour over a one- or three-year term, organizations can unlock potential savings (Lucidity Cloud, 2025). Azure and GCP have comparable offerings: Azure Reservations and GCP Commitment Matching (Microsoft, n.d.; Google, n.d.).



### 2.3 Spot Instances



Spot Instances allow users to bid on unused EC2 capacity at up to an 80% discount compared to On-Demand prices. This strategy is suitable for fault-tolerant and flexible workloads (AWS, n.d.). Azure and GCP offer similar spot instance models: Azure Low-Priority VMs and Google Preemptible VMs (Microsoft, n.d.; Google, n.d.).



### 2.4 Cost Allocation Tags



Implementing cost allocation tags enables organizations to track and categorize cloud spending by department, project, or other relevant criteria. This facilitates better budget management and cost visibility across the enterprise (Lucidity Cloud, 2025).



## 3. Workload Optimization Techniques



### 3.1 Rightsizing Instances



Regularly monitoring usage patterns and performance metrics allows organizations to rightsize instances, avoiding overprovisioning and unnecessary costs (Onescience Research, n.d.). Automated tools like AWS Trusted Advisor and Azure Advisor can assist in identifying underutilized resources (AWS, n.d.; Microsoft, n.d.).



### 3.2 Auto-Scaling



Auto-scaling enables organizations to adjust resource levels based on demand, ensuring optimal performance during peak times while minimizing costs during off-peak hours (KnowledgeHut, 2024). All major CSPs support auto-scaling features (AWS, n.d.; Microsoft, n.d.; Google, n.d.).



### 3.3 Storage Optimization



Optimizing storage utilization through techniques like data compression, archival storage, and lifecycle management policies can significantly reduce costs (Lucidity Cloud, 2025). AWS S3 Intelligent-Tiering and Azure Blob Storage Lifecycle Management are examples of such tools (AWS, n.d.; Microsoft, n.d.).



## 4. Case Studies: Successful Cost Reduction



### 4.1 Netflix



Netflix migrated its infrastructure to AWS, leveraging a multi-cloud strategy and rightsizing resources to support millions of users globally (Fynd Academy, 2025). By implementing cost optimization best practices, Netflix reduced its cloud spending by 60% (MoldStud, 2025).



### 4.2 Airbnb



Airbnb adopted a multi-cloud strategy using AWS, Azure, and GCP to optimize costs and ensure business continuity. By implementing rightsizing, storage optimization, and automated provisioning tools, Airbnb achieved a 60% reduction in cloud spending (EsparkInfo, 2024).



### 4.3 Accenture



Accenture utilized a multi-cloud approach for its Minna Bank project, building Japan's first digital bank on AWS. By implementing cost allocation tags, auto-scaling, and reserved instances, Accenture reduced cloud costs by 50% (Accenture, n.d.).



## 5. Conclusion



Optimizing cloud computing costs for large enterprises requires a strategic approach that considers pricing models, workload optimization techniques, and real-world case studies. By implementing best practices such as reserved instances, savings plans, spot instances, cost allocation tags, rightsizing instances, auto-scaling, and storage optimization, organizations can significantly reduce expenses while maintaining performance, scalability, and security. Successful cost reduction examples from Netflix, Airbnb, and Accenture demonstrate the tangible benefits of adopting these strategies across major CSPs.



## 6. References



* Accenture. (n.d.). *Cloud case studies*. [url website](https://www.accenture.com/us-en/about/company/cloud-case-studies)

* Amazon Web Services (AWS). (n.d.). *Reserved Instances*. [url website](https://aws.amazon.com/ec2/reserved-instances/)

* Amazon Web Services (AWS). (n.d.). *Spot Instances*. [url website](https://aws.amazon.com/ec2/spot/)

* Azure. (n.d.). *Azure Reserved Virtual Machine Instances*. [url website](https://azure.microsoft.com/en-us/pricing/reserved-virtual-machines/)

* Azure. (n.d.). *Low-priority VMs*. [url website](https://docs.microsoft.com/en-us/azure/virtual-machines/low-priority)

* EsparkInfo. (2024). *Cloud Cost Optimization: The Ultimate Guide for 2024*. [url website](https://www.esparkinfo.com/blog/cloud-cost-optimization)

* Fynd Academy. (2025). *7 Cloud Computing Case Study & Success Stories 2025*. [url website](https://www.fynd.academy/blog/cloud-computing-case-study)

* Google. (n.d.). *Commitment Matching*. [url website](https://cloud.google.com/commitment-matching)

* KnowledgeHut. (2024). *Cloud Computing Case Studies and Success Stories 2024*. [url website](https://www.knowledgehut.com/blog/cloud-computing/cloud-computing-case-studies)

* Lucidity Cloud. (2025). *13 Cloud Cost Optimization Best Practices*. [url website](https://www.lucidity.cloud/blog/cloud-cost-optimization-best-practices)

* Microsoft. (n.d.). *Azure Reservations*. [url website](https://azure.microsoft.com/en-us/pricing/reserved-virtual-machines/)

* MoldStud. (2025). *Successful Cloud Computing Implementations in Real Cases*. [url website](https://moldstud.com/articles/p-real-world-case-studies-successful-implementations-of-cloud-computing-models)

* Onescience Research. (n.d.). *Cost Optimization Strategies for Cloud Infrastructure*. [url website](https://www.onlinescientificresearch.com/articles/cost-optimization-strategies-for-cloud-infrastructure.html)

</details>

<details>
<summary>Who is Gabe Goodhart at IBM?</summary>

# Gabe Goodhart at IBM: A Comprehensive Analysis



## Introduction



Gabe Goodhart (LinkedIn handle: gabe-l-hart, GitHub: [gabe-l-hart](https://github.com/gabe-l-hart)) is a prominent figure in the realm of Artificial Intelligence (AI) and software development at IBM. His role as Software Architect and Deputy Chief Architect for the watsonx Platform underscores his significant contributions to IBM's AI initiatives ([IBM, 2025]). This report delves into Goodhart's professional background, achievements, and notable projects, providing an in-depth understanding of his role at IBM.



## Professional Background



Goodhart is a Software Architect and Deputy Chief Architect for the watsonx Platform at IBM ([LinkedIn, 2025a]). His expertise lies in AI, software development, and open-source technologies. He holds a position of influence within IBM's AI ecosystem, contributing to the company's strategic direction in AI research and development.



## Achievements and Projects



### Open Sourcing Granite Models



One of Goodhart's most notable achievements is his involvement in open-sourcing IBM's Granite series of Large Language Models (LLMs) ([LinkedIn, 2024b]). In May 2024, he announced the release of these models under a permissive Apache 2.0 license ([IBM Developer, 2024]). This move signified IBM's commitment to open AI and fostered collaboration within the developer community.



Goodhart has since shared detailed tutorials on leveraging Granite Code for local AI co-pilot setups using Visual Studio Code, Ollama, and Continue ([LinkedIn, 2024a]). These guides aim to help developers overcome common enterprise challenges such as data privacy, licensing, and cost.



### Collaboration with Open Source Community



Goodhart actively collaborates with various open-source projects and communities. For instance, he worked closely with the Ollama team to integrate Granite Code models seamlessly across multiple platforms ([LinkedIn, 2024c]). This collaboration highlights his dedication to fostering a robust AI ecosystem that benefits all developers.



### All Things Open Conference Contributions



Goodhart has presented at the All Things Open Conference, sharing insights on extending local AI developer toolsets using open-source tools like Ollama, Continue, and VS Code ([All Things Open Conference, 2024]). His contributions to this conference demonstrate his commitment to advancing AI development practices within the open-source community.



## Thought Leadership



Goodhart's thought leadership extends beyond his role at IBM. He frequently shares insights on AI and software development trends via LinkedIn posts, Twitter updates, and Medium articles ([Gabe Goodhart, 2024]). His expertise in AI-driven developer environments has garnered significant attention within the tech community.



## Conclusion



Gabe Goodhart's role at IBM is characterized by his contributions to open-source AI initiatives and collaborations with various communities. His work on open-sourcing Granite models and developing local AI co-pilot setups has significantly impacted developer practices, promoting transparency, cost-efficiency, and scalability in AI applications.



Goodhart's commitment to fostering a robust AI ecosystem aligns with IBM's broader strategy of shaping an innovative, collaborative, effective, and safe future for AI. His thought leadership and active engagement with the tech community further solidify his position as a key figure in IBM's AI landscape.



## References



**IBM.** (2025). *Gabe Goodhart | GitHub*. [GitHub](https://github.com/gabe-l-hart)



**LinkedIn.** (2024a). *We're excited to launch Granite 3.0! I particularly want to thank Michael…* [LinkedIn](https://www.linkedin.com/posts/gabe-goodhart-6676ba30_were-excited-to-launch-granite-30-i-particularly-activity-7254134037127049218-5GsS)



**LinkedIn.** (2024b). *In May, IBM and Red Hat announced that we were open sourcing our Granite…* [LinkedIn](https://www.linkedin.com/posts/gabe-goodhart-6676ba30_in-may-ibm-and-red-hat-announced-that-we-activity-7226681900520681473-R_mb)



**LinkedIn.** (2024c). *New on We ❤️ Open Source* [LinkedIn](https://www.linkedin.com/posts/all-things-open_new-on-we-open-source-the-journey-activity-7229124888018518018-K6LX)



**All Things Open Conference.** (2024). *#WeLoveOpenSource #AI #DevOps #opensource #softwaredevelopment…* [All Things Open](https://www.allthingsopen.org/)



**IBM Developer.** (2024). *Developing a gen AI application using IBM Granite Code*. [Medium](https://medium.com/@IBMDeveloper/developing-a-gen-ai-application-using-ibm-granite-code-18d206057158)



**Gabe Goodhart.** (2024). *Twitter Updates*. [Twitter](https://twitter.com/gabe_l_hart)

</details>

assafelovic commented:
Thank you @gabe-l-hart this is great! A lot has changed in the codebase since then, would you mind merging and resolving conflicts?

assafelovic commented:
Thank you @gabe-l-hart one last request sorry for not mentioning earlier - if you can update the docs relating updated changes. For example you can check out it here: https://github.com/assafelovic/gpt-researcher/blob/master/docs/docs/gpt-researcher/llms/llms.md

There might be more occasions in docs worth updating like changes to config if needed: https://github.com/assafelovic/gpt-researcher/blob/master/docs/docs/gpt-researcher/gptr/config.md

Thanks and happy to merge!

gabe-l-hart commented:
Thanks @assafelovic! I've updated both docs locations. I wasn't quite sure how to handle the framing of Granite in the LLM section since it runs under Ollama, so I added a new sub-section (`###`) indicating how to get the best use out of Granite under the main `## Ollama` section. Please let me know if there's a different way you'd like this to show up!

Also, we _just_ released `granite3.3` last week, and in the process changed how documents are concatenated (again), so I extended the `PromptFamily` to handle all `3.X` flavors of Granite.

assafelovic reviewed (approved):
This is super awesome and love your hard work around it. Thank you and excited to merge this! Feel free to share with IBM, etc

gabe-l-hart commented:
Thank you @assafelovic! I'm really excited to get this in. We're big fans of your project and have been using it a lot internally.
